### ISSUE TITLE
"Z-stack" container

### ISSUE THREAD

**Issue by birktj** [typst/typst#682]
It would be useful to have a container where elements would not influence the position of subsequent elements, but would still grow the size of the container. Eg. they would be stacked in the "z-dimension". We can use place to position elements without affecting positioning of other elements, but then they don't grow the size of the parent container either. This would be useful for various graphics 

Consider the following code that illustrates how I would imagine that such a container works:
```
#place(line(end: (30pt, 30pt)))
#place(line(start: (0pt, 30pt), end: (30pt, 0pt)))
Hello

#v(30pt)

#zstack(
    line(end: (30pt, 30pt)),
    line(start: (0pt, 30pt), end: (30pt, 0pt)),
)
Hello
```
![image](https://user-images.githubusercontent.com/2932651/230742727-63d060e7-745a-4634-95de-1ef8bba59ee6.png)

`zstack` can be implemented in Typst today as follows, but I think that it potentially useful enough to be included in the standard library:
```
#let zstack(..args) = style(styles => {
    let width = 0pt
    let height = 0pt
    for item in args.pos() {
        let size = measure(item, styles)
        width = calc.max(width, size.width)
        height = calc.max(height, size.height)
    }
    block(width: width, height: height, {
        for item in args.pos() {
            place(item)
        }
    })
})
```

**Comment by laurmaedje** [comment]
I agree that this is useful. I had thought about having this years ago, but never came around to just implement it (even though its probably pretty simple.) About the name: What do you think about `overlay` instead? Might be a bit beginner-friendly.

**Comment by laurmaedje** [comment]
> By the way, should this overlay element be able to be inlined, just like box?

What do you mean? To make it inline, you should wrap it into a box. In that sense, it should behave like most other elements (stack, grid, rotate, etc.)

**Comment by PgBiel** [comment]
> What do you mean? To make it inline, you should wrap it in a box. In that sense, it should behave like most other elements (stack, grid, rotate, etc.)

Ohh, I didn't consider that; thanks for the insight. It's mostly because I was thinking of designing overlay or layering functionality in general around the idea of adding stuff on top of existing content (a "base"), but I didn't think a box would just work (at least in text mode, but it's probably not that simple in math mode, as boxes can change the layout there - I wonder if this should be considered a bug?).

**Comment by laurmaedje** [comment]
The interaction of math and normal layouts is totally broken right now, but I plan to improve it.

**Comment by ModProg** [comment]
Maybe related, I would like to see a container that can inherit the size of its first element e.g., an image and allow other elements to use relative positions e.g., `center+horizon` for text overlays.

I guess one could implement this already using `measure`. But it would be great to have something like this built-in.

**Comment by PgBiel** [comment]
> Maybe related, I would like to see a container that can inherit the size of its first element e.g., an image and allow other elements to use relative positions e.g., `center+horizon` for text overlays.
> 
> I guess one could implement this already using `measure`. But it would be great to have something like this built-in.

Yes, this is more or less what I achieved with [my test implementation](https://github.com/typst/typst/compare/main...PgBiel:typst:overlay-tests).
Ideally, however, you'd be able to do something like `#overlay(a, b, c, d, e)` and things would be pretty much automatic, with the largest element defining the total size of the overlay etc. I don't think this is nearly as simple to do though, and just taking the first element as a base (as I did) is much simpler to do in the current infrastructure. Still, it'd be nice to discuss this further, with possible approaches for the ideal way to do it.

**Comment by ramkumarkb** [comment]
@birktj @PgBiel  - Would this be similar to the implementation described here - https://wiki.contextgarden.net/Layers 

Essentially, one defines a Layer and then puts some content in it. And `Layer` does not meddle with the actual typesetted content that is `behind`. And then comes some tricky parts to switch on/off the layers based on some conditions (which I assume can be done by a function).

Just some of my thoughts.

**Comment by stellarpower** [comment]
I agree that "layers" is the name it is given in most end user-facing applications. I believe Z-order is familiar for web developers however.

Bit of an aside, but I have been thinking about how an MS Publisher-style GUI could at some point really help broaden the userbase of Typst. It's well-designed and so I think it's plausible that a GUI beyond the web editor where I can click on things change their properties, lay them out, etc., could simultaneously update the source text, giving us the best of both worlds.

The idea of laying things on top of each other is well-etched into standard desktop metaphor for non-technical users. And my first thoughts are to "this gets rid of LaTex", and thus technical users like me - but - I have a couple of friends who have done some sort of publishing in their careers and I can see Typst as a viable tool for them down the line. Being able to stack things in the Z-axis would be great - I arrived here as I just wanted to add an image in the very top corner and so ignore the page margins for that image only (done with header-ascent) - but also, separating elements out into contextual layers (as one would do in GIMP, Inkscape, LibreOffice etc.) may prove to be a good tool for helping people design their documents in a way that doesn't end up being brittle and transition from more WYSIWYG tools.

**Comment by laurmaedje** [comment]
Fully agreed! More GUI-like tools are a thing that's on our radar, but probably quite far down the road.

**Comment by alkeryn** [comment]
wouldn't it make sense to add a dz option in move() ?

**Comment by laurmaedje** [comment]
I think a Z-stack container is overall more useful. I also ran into the lack of it a couple of times now and it's not only about the layering, but also about having things align properly.

**Comment by Lypsilonx** [comment]
I use this as a workaround for now:
```
#let z-stack(..items) = {
  grid(
    columns: items.pos().len() * (1fr,),
    column-gutter: -100%,
    rows: 1,
    ..items
  )
}
```

**Comment by SylvanFranklin** [comment]
Is there any progress on this? I'm trying to do something similar

**Comment by Risiamu** [comment]
I believe this feature is extremely important for less "standardized" typesetting work (I’m referring to projects beyond professional books and academic papers, such as TRPG rulebooks, art collections, magazines, etc., which typically fall under InDesign’s domain). While it’s currently possible to achieve similar results by adjusting the order of command execution, there still feels like something is lacking.

**Comment by Risiamu** [comment]
As someone who engages in both academic/professional book typesetting and spends significant time working with InDesign, I prefer the idea proposed in https://github.com/typst/typst/issues/2603. This approach would allow more flexible placement of images at varying positions as text flows.  
(Perhaps the `z` in #2603 should support positive/negative values, where `z=0` refers to all elements without an explicitly defined z-index.)

**Comment by Risiamu** [comment]
![Image](https://github.com/user-attachments/assets/4068bfed-d9d5-41dd-b99b-94340c5a1b09)
This is a great example. (Blurred for Copyright protection)
Currently, I can only observe the position of the text and then... 

```typ
#let image-place(path, dx: 0pt, dy: 0pt, width: auto) = {
  place(
    center + horizon,
    dx: dx,
    dy: dy,
    [#box(image(path, width: width))]
  )
}

#image-place("assets/2.JPG", dx: -10%, dy: 5%, width: 85%)
// before all the text start
```
To achieve this overlay effect.

It would be much more flexible if we can just:

```typ
#place(z: -1,[#image])
```

**Comment by philocalyst** [comment]
Would love this!
